This patch provides Mac-only extensions to GNU Emacs 22 CVS (Carbon Emacs).
Contributors are as follows.
mac_show_menu_bar -- S. Zenitani & T. Hiromatsu
mac_hide_menu_bar -- S. Zenitani & T. Hiromatsu
mac_dialog           -- S. Zenitani
mac_dialog_y_or_n_p  -- S. Zenitani
mac_spotlight_search -- S. Zenitani
mac_process_activate -- S. Zenitani
*** src/macfns.c	29 Mar 2008 00:45:58 -0000	1.106.2.6
--- src/macfns.c	1 Apr 2008 07:07:03 -0000
***************
*** 4588,4593 ****
--- 4590,4759 ----
    UNBLOCK_INPUT;
    return result;
  }
+ 
+ #if MAC_OSX 
+ 
+ DEFUN ("mac-show-menu-bar", Fmac_show_menu_bar, Smac_show_menu_bar, 0, 0, "",
+        doc: /* Show the menu bar.  */)
+     ()
+ {
+   if ( EQ (Vwindow_system, intern ("mac")))
+     ShowMenuBar();
+   return Qnil;
+ }
+ DEFUN ("mac-hide-menu-bar", Fmac_hide_menu_bar, Smac_hide_menu_bar, 0, 0, "",
+        doc: /* Hide the menu bar.  */)
+     ()
+ {
+   if ( EQ (Vwindow_system, intern ("mac")))
+     HideMenuBar();
+   return Qnil;
+ }
+ DEFUN ("mac-dialog", Fmac_dialog, Smac_dialog, 2, 2, "",
+        doc: /* Show a dialog.  */)
+   (message, explanation)
+      Lisp_Object message, explanation;
+ {
+   DialogRef           sheet=NULL;
+   AlertStdCFStringAlertParamRec alertParams;
+   struct frame *f = XFRAME (selected_frame);
+   CHECK_STRING (message);
+   CHECK_STRING (explanation);
+ 
+   if (FRAME_MAC_WINDOW (f))
+     {
+       GetStandardAlertDefaultParams(&alertParams,kStdCFStringAlertVersionOne);
+       alertParams.defaultText=(CFStringRef)kAlertDefaultOKText;       
+       alertParams.defaultButton=kAlertStdAlertOKButton;       
+ 
+       BLOCK_INPUT;
+       {
+         CFStringRef msgTitle = NULL, msgExplanation = NULL;
+         msgTitle       = cfstring_create_with_string (message);
+         msgExplanation = cfstring_create_with_string (explanation);
+ 
+         OSStatus err;
+         err = CreateStandardAlert(kAlertNoteAlert,
+                                   msgTitle,
+                                   msgExplanation,
+                                   &alertParams,
+                                   &sheet);
+         if(err == noErr)
+           err = RunStandardAlert(sheet,
+                                  NULL,
+                                  NULL);
+         CFRelease (msgTitle);
+         CFRelease (msgExplanation);
+       }
+       UNBLOCK_INPUT;
+     }
+   return Qnil;
+ }
+ DEFUN ("mac-dialog-y-or-n-p", Fmac_dialog_y_or_n_p, Smac_dialog_y_or_n_p, 2, 2, "",
+        doc: /* Show a dialog, which asks user a "Yes or No" question.
+ Return t if answer is "Yes". Takes two arguments,
+ which is the strings to display to ask the question.  */)
+   (message, explanation)
+      Lisp_Object message, explanation;
+ {
+   DialogRef           sheet=NULL;
+   AlertStdCFStringAlertParamRec alertParams;
+   struct frame *f = XFRAME (selected_frame);
+   CHECK_STRING (message);
+   CHECK_STRING (explanation);
+   SInt16 index = 0;
+ 
+   if (FRAME_MAC_WINDOW (f))
+     {
+       GetStandardAlertDefaultParams(&alertParams,kStdCFStringAlertVersionOne);
+       alertParams.defaultText=CFSTR ("Yes");
+       alertParams.defaultButton=kAlertStdAlertOKButton;       
+       alertParams.cancelText=CFSTR ("No");
+       alertParams.cancelButton=kAlertStdAlertCancelButton;
+ 
+       BLOCK_INPUT;
+       {
+         CFStringRef msgTitle = NULL, msgExplanation = NULL;
+         msgTitle       = cfstring_create_with_string (message);
+         msgExplanation = cfstring_create_with_string (explanation);
+ 
+         OSStatus err;
+         err = CreateStandardAlert(kAlertNoteAlert,
+                                   msgTitle,
+                                   msgExplanation,
+                                   &alertParams,
+                                   &sheet);
+         if(err == noErr)
+           err = RunStandardAlert(sheet,
+                                  NULL,
+                                  &index);
+         CFRelease (msgTitle);
+         CFRelease (msgExplanation);
+       }
+       UNBLOCK_INPUT;
+     }
+   if( index==1 ){
+     return Qt;
+   }else{
+     return Qnil;
+   }
+ }
+ DEFUN ("mac-spotlight-search", Fmac_spotlight_search,
+        Smac_spotlight_search, 1, 1, "MSearch in Spotlight: ",
+        doc: /* Search STRING with Spotlight. */)
+      (string)
+      Lisp_Object string;
+ {
+   CFStringRef search_str;
+   search_str = cfstring_create_with_string ( string );
+   if( search_str != NULL )
+     HISearchWindowShow(search_str,kNilOptions);
+   CFRelease (search_str);
+   return Qnil;
+ }
+ /* Originally from Kurita-san's SmartActivate (Cocoa) */
+ DEFUN ("mac-process-activate", Fmac_process_activate,
+        Smac_process_activate, 1, 1, 0,
+        doc: /* Show up the frontmost window of a Mac OSX application process.
+ Pass a bundle identifier IDENTIFIER to specify the application. */)
+      (identifier)
+      Lisp_Object identifier;
+ {
+   CFDictionaryRef pDict = NULL;
+   CFStringRef pDictKey = CFSTR( "CFBundleIdentifier" ), idKey = NULL;
+   ProcessSerialNumber psn = {kNoProcess, kNoProcess};
+   Boolean isFound = false;
+ 
+   if (STRINGP (identifier))
+     idKey = cfstring_create_with_string (identifier);
+   else
+     return Qnil;
+ 
+   OSErr err = GetNextProcess (&psn);
+   while( err == noErr)
+     {
+       pDict = ProcessInformationCopyDictionary
+       (&psn, kProcessDictionaryIncludeAllInformationMask);
+       if( CFDictionaryContainsKey( pDict, pDictKey ) )
+         {
+           CFStringRef dictValue = CFDictionaryGetValue( pDict, pDictKey );
+           if (dictValue != NULL)
+             if( CFStringCompare ( dictValue, idKey, 0 ) == 0 )
+               isFound = true;
+         }
+       CFRelease( pDict );
+       if (isFound) break;
+       err = GetNextProcess (&psn);
+     }
+ 
+   if( isFound )
+     SetFrontProcessWithOptions ( &psn, kSetFrontProcessFrontWindowOnly );
+ 
+   return Qnil;
+ }
+ 
+ #endif /*MAC_OSX*/
+ 
  #endif
  
  
***************
*** 4887,4892 ****
--- 5419,5432 ----
  
  #if TARGET_API_MAC_CARBON
    defsubr (&Sx_file_dialog);
+ #if MAC_OSX
+   defsubr (&Smac_show_menu_bar);
+   defsubr (&Smac_hide_menu_bar);
+   defsubr (&Smac_dialog);
+   defsubr (&Smac_dialog_y_or_n_p);
+   defsubr (&Smac_spotlight_search);
+   defsubr (&Smac_process_activate);
+ #endif
  #endif
    defsubr (&Smac_clear_font_name_table);
  #if USE_MAC_FONT_PANEL
